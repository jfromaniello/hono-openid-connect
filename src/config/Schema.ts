import { CookieStore } from "hono-sessions";
import { z } from "zod";

const isHttps = /^https:/i;

// Helper method to create response type/mode schema with proper dependencies
const createAuthParamsSchema = () => {
  return z
    .object({
      response_type: z
        .enum(["id_token", "code id_token", "code"])
        .optional()
        .default("code"),
      scope: z
        .string()
        .regex(/\bopenid\b/, "Must contain openid")
        .optional()
        .default("openid profile email"),
      response_mode: z.string().optional(),
    })
    .passthrough();
};

// Create the configuration schema
export const ConfigurationSchema = z
  .object({
    session: z.union([
      z.literal(false),
      z
        .object({
          store: z.any().optional().default(new CookieStore()),
          encryptionKey: z.string().min(32),
          expireAfterSeconds: z
            .number()
            .optional()
            .default(60 * 60 * 24),
          sessionCookieName: z.string().optional().default("appSession"),
          cookieOptions: z
            .object({
              domain: z.string().optional(),
              sameSite: z
                .enum(["Lax", "Strict", "None"])
                .optional()
                .default("Lax"),
              path: z.string().optional().default("/"),
              httpOnly: z.boolean().optional().default(true),
              secure: z.boolean().optional(),
              maxAge: z.number().optional(),
              sessionCookieName: z.string().optional().default("session"),
            })
            .optional()
            .default({}),
        })
        .required(),
    ]),
    tokenEndpointParams: z.record(z.any()).optional(),
    authorizationParams: createAuthParamsSchema().optional().default({}),
    logoutParams: z.record(z.any()).optional(),
    baseURL: z.string().url(),
    clientID: z.string(),
    clientSecret: z.string().optional(),
    clockTolerance: z.number().optional().default(60),
    enableTelemetry: z.boolean().optional().default(true),
    errorOnRequiredAuth: z.boolean().optional().default(false),
    attemptSilentLogin: z.boolean().optional().default(false),
    excludedClaims: z
      .array(z.string())
      .optional()
      .default([
        "aud",
        "iss",
        "iat",
        "exp",
        "nbf",
        "nonce",
        "azp",
        "auth_time",
        "s_hash",
        "at_hash",
        "c_hash",
      ]),
    idpLogout: z.boolean().optional().default(false),
    idTokenSigningAlg: z
      .string()
      .refine((val) => val.toLowerCase() !== "none", {
        message: "Signing algorithm cannot be 'none'",
      })
      .optional()
      .default("RS256"),
    issuerBaseURL: z.string().url(),
    authRequired: z.boolean().optional().default(true),
    pushedAuthorizationRequests: z.boolean().optional().default(false),
    customRoutes: z
      .array(z.enum(["login", "callback", "logout"]))
      .optional()
      .default([]),
    debug: z
      .custom<(message: string, metadata?: Record<string, unknown>) => void>(
        (v) => typeof v === "function",
      )
      .optional()
      .default(() => () => {}),
    routes: z
      .object({
        login: z.string().regex(/^\//).optional().default("/login"),
        logout: z.string().regex(/^\//).optional().default("/logout"),
        callback: z.string().regex(/^\//).optional().default("/callback"),
      })
      .optional()
      .default({}),
    clientAuthMethod: z
      .enum([
        "client_secret_basic",
        "client_secret_post",
        "client_secret_jwt",
        "private_key_jwt",
        "none",
      ])
      .optional(),
    clientAssertionSigningKey: z.any().optional(),
    clientAssertionSigningAlg: z
      .enum([
        "RS256",
        "RS384",
        "RS512",
        "PS256",
        "PS384",
        "PS512",
        "ES256",
        "ES256K",
        "ES384",
        "ES512",
        "EdDSA",
      ])
      .optional(),
    discoveryCacheMaxAge: z
      .number()
      .min(0)
      .optional()
      .default(10 * 60 * 1000),
    httpTimeout: z.number().min(500).optional().default(5000),
    httpUserAgent: z.string().optional().default("hono-openid-connect"),

    fetch: z
      .custom<typeof globalThis.fetch>((v) => typeof v === "function")
      .optional()
      .default(() => globalThis.fetch),
  })
  .superRefine((data, ctx) => {
    // Handle secure cookie validation based on baseURL
    if (data.session && typeof data.session !== "boolean") {
      const cookieOptions = data.session.cookieOptions;
      if (isHttps.test(data.baseURL)) {
        if (cookieOptions.secure === false) {
          ctx.addIssue({
            code: z.ZodIssueCode.custom,
            message:
              "Setting your cookie to insecure when over https is not recommended, I hope you know what you're doing.",
            path: ["session", "cookieOptions", "secure"],
          });
        } else if (cookieOptions.secure === undefined) {
          // Default to true for HTTPS
          data.session.cookieOptions.secure = true;
        }
      } else if (cookieOptions.secure === true) {
        // Error for HTTP with secure cookie
        ctx.addIssue({
          code: z.ZodIssueCode.custom,
          message:
            "Cookies set with the `Secure` property won't be attached to http requests",
          path: ["session", "cookieOptions", "secure"],
        });
      } else if (cookieOptions.secure === undefined) {
        // Default to false for HTTP
        data.session.cookieOptions.secure = false;
      }
    }

    // Validate response_mode based on response_type
    if (data.authorizationParams) {
      const responseType = data.authorizationParams.response_type;
      const responseMode = data.authorizationParams.response_mode;

      if (responseType === "code") {
        if (responseMode && !["query", "form_post"].includes(responseMode)) {
          ctx.addIssue({
            code: z.ZodIssueCode.invalid_enum_value,
            options: ["query", "form_post"],
            received: responseMode,
            message:
              "For response_type 'code', response_mode must be 'query' or 'form_post'",
            path: ["authorizationParams", "response_mode"],
          });
        }
      } else if (responseMode && responseMode !== "form_post") {
        ctx.addIssue({
          code: z.ZodIssueCode.invalid_enum_value,
          options: ["form_post"],
          received: responseMode,
          message: "For this response_type, response_mode must be 'form_post'",
          path: ["authorizationParams", "response_mode"],
        });
      } else if (!responseMode) {
        // Set default for non-code response types
        data.authorizationParams.response_mode = "form_post";
      }

      // Warning about form_post with HTTP baseURL
      if (
        data.authorizationParams.response_mode === "form_post" &&
        !isHttps.test(data.baseURL)
      ) {
        ctx.addIssue({
          code: z.ZodIssueCode.custom,
          message:
            "Using 'form_post' for response_mode may cause issues for you logging in over http, see https://github.com/auth0/express-openid-connect/blob/master/FAQ.md",
          path: ["baseURL"],
        });
      }
    }

    // Validate clientSecret requirements
    if (
      data.clientAuthMethod &&
      data.clientAuthMethod.includes("client_secret") &&
      !data.clientSecret
    ) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: `"clientSecret" is required for the "clientAuthMethod" "${data.clientAuthMethod}"`,
        path: ["clientSecret"],
      });
    }

    if (
      data.idTokenSigningAlg &&
      data.idTokenSigningAlg.startsWith("HS") &&
      !data.clientSecret
    ) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message:
          '"clientSecret" is required for ID tokens with HMAC based algorithms',
        path: ["clientSecret"],
      });
    }

    // Validate clientAuthMethod
    if (
      data.authorizationParams?.response_type?.includes("code") &&
      data.clientAuthMethod === "none"
    ) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: "Public code flow clients are not supported.",
        path: ["clientAuthMethod"],
      });
    }

    if (data.pushedAuthorizationRequests && data.clientAuthMethod === "none") {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: "Public PAR clients are not supported.",
        path: ["clientAuthMethod"],
      });
    }

    // Set default clientAuthMethod
    if (data.clientAuthMethod === undefined) {
      if (
        data.authorizationParams?.response_type === "id_token" &&
        !data.pushedAuthorizationRequests
      ) {
        data.clientAuthMethod = "none";
      } else if (data.clientAssertionSigningKey) {
        data.clientAuthMethod = "private_key_jwt";
      } else {
        data.clientAuthMethod = "client_secret_basic";
      }
    }

    // Validate clientAssertionSigningKey
    if (
      data.clientAuthMethod === "private_key_jwt" &&
      !data.clientAssertionSigningKey
    ) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message:
          '"clientAssertionSigningKey" is required for a "clientAuthMethod" of "private_key_jwt"',
        path: ["clientAssertionSigningKey"],
      });
    }
  });
