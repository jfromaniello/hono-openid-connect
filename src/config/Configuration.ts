import { OIDCAuthorizationRequestParams } from "@/config/authRequest.js";
import { SessionOptions } from "@jfromaniello/hono-sessions";

type Routes = {
  login: string;
  logout: string;
  callback: string;
};

export interface Configuration {
  /**
   * The base URL of the OIDC provider.
   */
  issuerBaseURL: string;

  /**
   * The base URL of the application.
   */
  baseURL: string;

  /**
   * The client ID of the application.
   */
  clientID: string;

  /**
   * The client secret of the application.
   */
  clientSecret?: string;

  /**
   * Whether to require authentication for all routes.
   * @default true
   */
  authRequired: boolean;

  /**
   * Whether to use the IDP's logout endpoint.
   * @default false
   */
  idpLogout: boolean;

  /**
   * hono-sessions options.
   *
   * @default {
   *  store: new CookieStore(),
   *  sessionCookieName: 'appSession',
   *  cookieOptions: {
   *    sameSite: 'Lax',
   *    path: '/',
   *    httpOnly: true,
   *  }
   * }
   */
  session: false | SessionOptions;

  /**
   * Use this setting to prevent the default routes from being installed.
   *
   * @default []
   */
  customRoutes: (keyof Routes)[];

  /**
   * Routes options.
   *
   * @default {
   *   login: '/login',
   *   logout: '/logout',
   *   callback: '/callback',
   * }
   */
  routes: Routes;

  /**
   * Additional authorization request parameters that will be included in
   * the authorization URL.
   *
   * @default {
   *  response_type: 'id_token',
   *  scope: 'openid profile email',
   *  response_mode: 'form_post',
   * }
   */
  authorizationParams: Partial<OIDCAuthorizationRequestParams>;

  /**
   * Additional parameters that will be sent to the
   * token endpoint, typically used for parameters such as `resource`
   * in cases where multiple resource indicators were requested but
   * the authorization server only supports issuing an access token
   * with a single audience
   */
  tokenEndpointParams?: Record<string, string>;

  /**
   * Whether to use pushed authorization requests.
   * @default false
   */
  pushedAuthorizationRequests: boolean;

  /**
   * The clock tolerance for the OIDC client.
   * @default 60
   */
  clockTolerance: number;

  /**
   * Whether to enable telemetry.
   * @default true
   */
  enableTelemetry: boolean;

  /**
   * The HTTP timeout for the OIDC client.
   * @default 5000
   */
  httpTimeout?: number;

  // Hooks
  // afterCallback?: (
  //   c: Context,
  //   session: OidcSession,
  //   state: any,
  // ) => Promise<OIDCUserInfoResponse> | OIDCUserInfoResponse;

  /**
   * The method to use for client authentication.
   *
   * If `authorizationParams.response_type` is `id_token` and !pushedAuthorizationRequests,
   * the @default is `none`.
   *
   * If `clientAssertionSigningKey` is provided,
   *  the @default is `private_key_jwt`.
   *
   * Otherwise, the @default is `client_secret_basic`.
   */
  clientAuthMethod:
    | "client_secret_basic"
    | "client_secret_post"
    | "client_secret_jwt"
    | "private_key_jwt"
    | "none";

  /**
   * The client assertion signing key.
   * Required if `clientAuthMethod` is `private_key_jwt`.
   */
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  clientAssertionSigningKey?: any;

  /**
   * The client assertion signing algorithm.
   * Required if `clientAuthMethod` is `private_key_jwt`.
   * @default 'RS256'
   */
  clientAssertionSigningAlg?:
    | "RS256"
    | "RS384"
    | "RS512"
    | "PS256"
    | "PS384"
    | "PS512"
    | "ES256"
    | "ES256K"
    | "ES384"
    | "ES512"
    | "EdDSA";

  /**
   * Returns 401 if the user is not authenticated.
   * @default false
   */
  errorOnRequiredAuth: boolean;

  /**
   * Whether to attempt a silent login.
   * @default false
   */
  attemptSilentLogin: boolean;

  /**
   * The claims to exclude from the user identity.
   * @default ['aud', 'iss', 'iat', 'exp', 'nbf', 'nonce', 'azp', 'auth_time', 's_hash', 'at_hash', 'c_hash']
   */
  excludedClaims: string[];

  /**
   * The expected signing algorithm for the ID token.
   * @default 'RS256'
   */
  idTokenSigningAlg: string;

  /**
   * The maximum age of the discovery cache.
   * @default 10 * 60 * 1000
   */
  discoveryCacheMaxAge: number;

  /**
   * The HTTP user agent to use for the OIDC client.
   * @default 'hono-openid-connect'
   */
  httpUserAgent: string;

  /**
   * The fetch function to use for the OIDC client.
   * @default globalThis.fetch
   */
  fetch: typeof globalThis.fetch;

  /**
   * Additional parameters that will be sent to the
   * logout endpoint.
   */
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  logoutParams?: Record<string, any>;

  /**
   * Logger function to use for the OIDC client.
   * @param message - The message to log.
   * @param metadata - Additional metadata to include in the log.
   */
  debug: (message: string, metadata?: Record<string, unknown>) => void;
}

// Type for the required fields that must be provided in InitConfiguration
type RequiredConfigFields = "issuerBaseURL" | "baseURL" | "clientID";

// Type for the optional session field that should be partial in InitConfiguration
type SessionField = {
  session?: false | Partial<SessionOptions>;
};

// Type for the optional routes field that should be partial in InitConfiguration
type RoutesField = {
  routes?: Partial<Routes>;
};

/**
 * Configuration type for initializing the OIDC client.
 * This represents the input before validation, where most properties are optional
 * since they have defaults in the schema.
 */
export type InitConfiguration = Pick<Configuration, RequiredConfigFields> &
  Partial<Omit<Configuration, RequiredConfigFields | "session" | "routes">> &
  SessionField &
  RoutesField;
