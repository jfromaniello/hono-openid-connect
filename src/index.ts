export { auth } from "./auth";

export { type UserInfoResponse as UserInfo } from "openid-client";
export type { OIDCEnv, OIDCVariables } from "./lib/honoEnv";
export { type OIDCSession } from "./types/session";

export {
  attemptSilentLogin,
  callback,
  login,
  logout,
  pauseSilentLogin,
  requiresAuth,
  resumeSilentLogin,
} from "./middleware";

export type { TokenEndpointResponse as TokenSet } from "openid-client";

export { OIDCException } from "./lib/Exception";
