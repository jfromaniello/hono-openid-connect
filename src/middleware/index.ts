export { callback } from "./callback.js";
export { login } from "./login.js";
export { logout } from "./logout.js";
export { requiresAuth } from "./requiresAuth.js";

//export all middlewares in this file
export {
  attemptSilentLogin,
  pauseSilentLogin,
  resumeSilentLogin,
} from "./silentLogin.js";
