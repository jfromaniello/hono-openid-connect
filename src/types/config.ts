import { Context } from 'hono';

// Extend the Hono context to include OIDC context
declare module 'hono' {
  interface ContextVariableMap {
    oidc: OidcContext;
    oidcConfig: OidcClientConfig;
  }
}

export interface OidcSession {
  user: UserinfoResponse;
  idToken: string;
  accessToken?: {
    token: string;
    token_type: string;
    expires_at: number;
    isExpired: () => boolean;
  };
  refreshToken?: string;
}

export interface OidcContext {
  isAuthenticated: boolean;
  user: UserinfoResponse;
  idToken: string;
  accessToken?: {
    token: string;
    token_type: string;
    expires_at: number;
    isExpired: () => boolean;
  };
  refreshToken?: string;
}

// For internal use
export interface OidcClientConfig {
  config: OidcConfig;
  oidcClient: any;
}

export interface LogoutToken {
  iss: string;
  sub: string;
  aud: string;
  iat: number;
  jti: string;
  events: {
    'http://schemas.openid.net/event/backchannel-logout': {};
  };
  sid?: string;
}
