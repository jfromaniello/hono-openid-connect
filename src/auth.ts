import { MiddlewareHandler, Next } from "hono";
import { sessionMiddleware } from "hono-sessions";
import { every } from "hono/combine";
import { createMiddleware } from "hono/factory";
import { HTTPException } from "hono/http-exception";
import { parseConfiguration } from "./config";
import { InitConfiguration } from "./config/Configuration";
import { initializeOidcClient } from "./lib/client";
import { OIDCContext } from "./lib/context";
import { OIDCEnv } from "./lib/honoEnv";
import {
  callback as callbackHandler,
  login as loginHandler,
  logout as logoutHandler,
} from "./middleware";

/**
 * Main auth middleware function
 */
export function auth(initConfig: InitConfiguration): MiddlewareHandler {
  const config = parseConfiguration(initConfig);

  // Initialize session middleware if needed
  const sessionMiddlewareHandler =
    config.session !== false ? sessionMiddleware(config.session) : null;

  // Main OIDC middleware function
  const oidcMiddleware: MiddlewareHandler = createMiddleware<OIDCEnv>(
    async (c, next: Next): Promise<Response | void> => {
      try {
        c.set("oidcConfiguration", config);

        // Initialize the OIDC client with defaults
        const clientConfig = await initializeOidcClient(config);
        c.set("oidcClient", clientConfig);

        // Check if there's a valid session
        const session = c.get("session");
        if (!session) {
          throw new HTTPException(500, {
            message: "Session middleware not configured properly",
          });
        }

        const oidcContext = new OIDCContext(c);
        c.set("oidc", oidcContext);

        // Use destructuring with defaults to ensure routes is always defined
        const { routes, authRequired } = config;
        const { login, callback, logout } = routes;

        // Handle login route
        if (!config.customRoutes.includes("login") && c.req.path === login) {
          return loginHandler()(c, next);
        }

        // Handle callback route
        if (
          !config.customRoutes.includes("callback") &&
          c.req.path === callback
        ) {
          return callbackHandler()(c, next);
        }

        // Handle logout route
        if (!config.customRoutes.includes("logout") && c.req.path === logout) {
          return logoutHandler()(c, next);
        }

        // Handle unauthenticated requests
        if (authRequired && !c.var.oidc?.isAuthenticated) {
          return loginHandler()(c, next);
        }
      } catch (error) {
        console.error("OIDC Middleware Error:", error);
        return c.text("Internal Server Error", 500);
      }

      // Continue to the next middleware or route handler
      return await next();
    },
  );

  // Return array of middlewares or just the OIDC middleware
  if (sessionMiddlewareHandler) {
    return every(sessionMiddlewareHandler, oidcMiddleware);
  }
  return oidcMiddleware;
}
