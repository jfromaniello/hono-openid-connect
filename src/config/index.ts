import { Context } from "hono";
import { HTTPException } from "hono/http-exception";
import { OIDCEnv } from "../lib/honoEnv";
import { Configuration, InitConfiguration } from "./Configuration";
import { ConfigurationSchema } from "./Schema";
// import { ConfigurationSchema } from "./Schema";
const parsedConfig = new Map<InitConfiguration, Configuration>();

export const parseConfiguration = (
  config: InitConfiguration,
): Configuration => {
  if (parsedConfig.has(config)) {
    return parsedConfig.get(config)!;
  }
  const result = ConfigurationSchema.parse(config) as Configuration;
  parsedConfig.set(config, result);
  return result;
};

export const getConfiguration = (c: Context<OIDCEnv>): Configuration => {
  if (!c.var.oidcConfiguration) {
    throw new HTTPException(500, {
      message:
        "The oidc middleware is not properly configured. Install `auth` first.",
    });
  }
  return c.var.oidcConfiguration;
};

export { assignFromEnv, type ConditionalInitConfig } from "./envConfig";
