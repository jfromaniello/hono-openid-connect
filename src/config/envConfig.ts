import { MakeOptional } from "../types/util";
import { InitConfiguration } from "./Configuration";

export type MinimalConfigByEnv = {
  OIDC_ISSUER_URL: string;
  OIDC_CLIENT_ID: string;
  OIDC_CLIENT_SECRET?: string;
  OIDC_AUDIENCE?: string;
  BASE_URL: string;
  OIDC_SESSION_ENCRYPTION_KEY?: string;
};

type PartialConfig = MakeOptional<
  InitConfiguration,
  "clientID" | "clientSecret" | "issuerBaseURL" | "baseURL" | "session"
>;

export type ConditionalInitConfig = NodeJS.ProcessEnv extends MinimalConfigByEnv
  ? PartialConfig
  : InitConfiguration;

export const envHasConfig = (
  config: MinimalConfigByEnv | unknown,
): config is MinimalConfigByEnv => {
  return (
    typeof config === "object" &&
    config !== null &&
    "OIDC_ISSUER_URL" in config &&
    "OIDC_CLIENT_ID" in config &&
    "BASE_URL" in config &&
    typeof config.OIDC_ISSUER_URL === "string" &&
    typeof config.OIDC_CLIENT_ID === "string" &&
    typeof config.BASE_URL === "string"
  );
};

export const assignFromEnv = (
  config: ConditionalInitConfig,
): InitConfiguration => {
  const configWithoutEnv = config ?? ({} as ConditionalInitConfig);

  if (!envHasConfig(process.env)) {
    return configWithoutEnv as InitConfiguration;
  }

  const {
    OIDC_ISSUER_URL,
    OIDC_CLIENT_ID,
    OIDC_CLIENT_SECRET,
    BASE_URL,
    OIDC_AUDIENCE,
  } = process.env;
  return {
    ...configWithoutEnv,
    issuerBaseURL: configWithoutEnv.issuerBaseURL ?? OIDC_ISSUER_URL,
    clientID: configWithoutEnv.clientID ?? OIDC_CLIENT_ID,
    clientSecret: configWithoutEnv.clientSecret ?? OIDC_CLIENT_SECRET,
    baseURL: configWithoutEnv.baseURL ?? BASE_URL,
    authorizationParams: OIDC_AUDIENCE
      ? { audience: OIDC_AUDIENCE }
      : undefined,
    session:
      configWithoutEnv.session === false
        ? false
        : {
            ...(configWithoutEnv.session || {}),
            encryptionKey:
              configWithoutEnv.session?.encryptionKey ??
              process.env.OIDC_SESSION_ENCRYPTION_KEY,
          },
  };
};
