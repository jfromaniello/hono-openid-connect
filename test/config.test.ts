/* eslint-disable @typescript-eslint/no-explicit-any */
import { describe, expect, it } from "vitest";
import { parseConfiguration } from "../src/config";
import { InitConfiguration } from "../src/config/Configuration";

describe("Configuration Parser", () => {
  it("should parse a valid configuration", () => {
    const validConfig: InitConfiguration = {
      issuerBaseURL: "https://auth.example.com",
      baseURL: "https://app.example.com",
      clientID: "test-client-id",
      clientSecret: "test",
      session: {
        encryptionKey: "test encryption key fdsgfds gfds ",
      },
    };

    const parsedConfig = parseConfiguration(validConfig);

    expect(parsedConfig).toHaveProperty(
      "issuerBaseURL",
      "https://auth.example.com",
    );
    expect(parsedConfig).toHaveProperty("baseURL", "https://app.example.com");
    expect(parsedConfig).toHaveProperty("clientID", "test-client-id");
    expect(parsedConfig).toHaveProperty("authRequired", true); // default value
    expect(parsedConfig).toHaveProperty("fetch", globalThis.fetch); // default value
    expect(parsedConfig).toMatchSnapshot();
  });

  it("should throw an error for invalid configuration", () => {
    const invalidConfig = {
      // Missing required issuerBaseURL
      baseURL: "https://app.example.com",
      clientID: "test-client-id",
    };

    expect(() => parseConfiguration(invalidConfig as any)).toThrow();
  });

  it("should apply default values to configuration", () => {
    const minimalConfig: InitConfiguration = {
      issuerBaseURL: "https://auth.example.com",
      baseURL: "https://app.example.com",
      clientID: "test-client-id",
      clientSecret: "test-client-secret",
      session: {
        encryptionKey: "test encryption key fdsgfds gfds ",
      },
    };

    const parsedConfig = parseConfiguration(minimalConfig);

    expect(parsedConfig).toHaveProperty("authRequired", true);
    expect(parsedConfig).toHaveProperty("clockTolerance", 60);
    expect(parsedConfig).toHaveProperty("routes");

    expect(parsedConfig.routes).toEqual({
      login: "/login",
      logout: "/logout",
      callback: "/callback",
    });
  });

  it("should cache parsed configurations", () => {
    const config: InitConfiguration = {
      issuerBaseURL: "https://auth.example.com",
      baseURL: "https://app.example.com",
      clientID: "test-client-id",
      clientSecret: "test",
      session: {
        encryptionKey: "test encryption key fdsgfds gfds ",
      },
    };

    const firstParsed = parseConfiguration(config);
    const secondParsed = parseConfiguration(config);

    expect(firstParsed).toBe(secondParsed); // Should be the same object instance (cached)
  });

  it("should not allow custom routes to be set to relative paths", () => {
    const config: InitConfiguration = {
      issuerBaseURL: "https://auth.example.com",
      baseURL: "https://app.example.com",
      clientID: "test-client-id",
      routes: {
        login: "login",
      },
      session: {
        encryptionKey: "test encryption key fdsgfds gfds ",
      },
    };

    expect(() => parseConfiguration(config)).toThrow();
  });
});
