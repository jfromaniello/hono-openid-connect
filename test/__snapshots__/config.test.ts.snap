// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Configuration Parser > should parse a valid configuration 1`] = `
{
  "attemptSilentLogin": false,
  "authRequired": true,
  "authorizationParams": {
    "response_type": "code",
    "scope": "openid profile email",
  },
  "baseURL": "https://app.example.com",
  "clientAuthMethod": "client_secret_basic",
  "clientID": "test-client-id",
  "clientSecret": "test",
  "clockTolerance": 60,
  "customRoutes": [],
  "debug": [Function],
  "discoveryCacheMaxAge": 600000,
  "enableTelemetry": true,
  "errorOnRequiredAuth": false,
  "excludedClaims": [
    "aud",
    "iss",
    "iat",
    "exp",
    "nbf",
    "nonce",
    "azp",
    "auth_time",
    "s_hash",
    "at_hash",
    "c_hash",
  ],
  "fetch": [Function],
  "httpTimeout": 5000,
  "httpUserAgent": "hono-openid-connect",
  "idTokenSigningAlg": "RS256",
  "idpLogout": false,
  "issuerBaseURL": "https://auth.example.com",
  "pushedAuthorizationRequests": false,
  "routes": {
    "callback": "/callback",
    "login": "/login",
    "logout": "/logout",
  },
  "session": {
    "cookieOptions": {
      "httpOnly": true,
      "path": "/",
      "sameSite": "Lax",
      "secure": true,
      "sessionCookieName": "session",
    },
    "encryptionKey": "test encryption key fdsgfds gfds ",
    "expireAfterSeconds": 86400,
    "sessionCookieName": "appSession",
    "store": CookieStore {
      "cookieOptions": undefined,
      "encryptionKey": undefined,
      "sessionCookieName": "session",
    },
  },
}
`;
